### PR TITLE
fix: allow type imports in moduleResolution: node

### LINKED ISSUE #7185: @typescript-eslint/utils imports no longer detect types in moduleResolution: node(10) in 6.0.0
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

After upgrading `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` to 6.0.0 (from 5.61.0) it appears that `@typescript-eslint/utils` can no longer be found, see Repro for details.

### Reproduction Repository Link

https://github.com/andreashuber69/eslint-config

### Repro Steps

```
git clone https://github.com/andreashuber69/eslint-config.git
cd eslint-config
npm i
npm run lint
```

The above steps produce no error (5.61.0 packages are installed). Now upgrade to 6.0.0 and lint again, as follows:

```
npm i @typescript-eslint/eslint-plugin@latest @typescript-eslint/parser@latest
npm run lint
```

I get the following errors:
```
/home/a/git/eslint-config/test/getSeverities.ts
   6:18  error  'any' overrides all other types in this union type  @typescript-eslint/no-redundant-type-constituents
   6:31  error  'any' overrides all other types in this union type  @typescript-eslint/no-redundant-type-constituents
  14:13  error  Unsafe return of an `any` typed value               @typescript-eslint/no-unsafe-return
  19:11  error  Unsafe assignment of an `any` value                 @typescript-eslint/no-unsafe-assignment
  29:13  error  Unsafe return of an `any` typed value               @typescript-eslint/no-unsafe-return

/home/a/git/eslint-config/test/typescriptEslintRules.ts
   5:69  error  Unsafe return of an `any` typed value       @typescript-eslint/no-unsafe-return
   5:81  error  Unsafe member access [0] on an `any` value  @typescript-eslint/no-unsafe-member-access
   8:11  error  Unsafe assignment of an `any` value         @typescript-eslint/no-unsafe-assignment
  14:5   error  Unsafe return of an `any` typed value       @typescript-eslint/no-unsafe-return
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.0.0` |
| `@typescript-eslint/parser`            | `6.0.0` |
| `@typescript-eslint/scope-manager`     | `6.0.0` |
| `@typescript-eslint/typescript-estree` | `6.0.0` |
| `@typescript-eslint/type-utils`        | `6.0.0` |
| `@typescript-eslint/utils`             | `6.0.0` |
| `TypeScript`                           | `5.1.6` |
| `ESLint`                               | `8.44.0` |
| `node`                                 | `18.16.1` |

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #7185
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Package types lookup in `"moduleResolution": "node"` adheres to Node 10 (`"moduleResolution": "node10"`), which doesn't support `"exports"` > `"."` > `"types"` field in the `package.json`. https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/a8dece2e3aab25f9ddbef1ac2f1c22ae8d27bc6f/docs/problems/NoResolution.md#true-positive-node-10-doesnt-support-packagejson-exports recommends using a newer `moduleResolution` in your TSConfig. To help out users still using Node 10 module resolution, this adds explicit `"types": "./(dist/)index.d.ts"` entries to the `package.json`s.

Note that you can avoid the need for this code change by setting a new module resolution strategy, such as `"moduleResolution": "node16"`, in your TSConfig:

```diff
{
  // node-lts-strictest is deprecated in favor of using an array to combine node-lts and strictest but doing so
  // currently leads to problems with the eslint-plugin-import package, see
  // https://github.com/import-js/eslint-plugin-import/issues/2751
  "extends": "@tsconfig/node-lts-strictest/tsconfig.json",
  "compilerOptions": {
+    "moduleResolution": "Node16",
  },
}

```

### CODE REVIEW

**Review by bradzacher — COMMENTED** [review]
This was an intentional removal though because we don't support node10 - we're now at a minimum of node16 - so supporting types for node10 resolution is wrong and allows people to do broken things (eg TS will happily let you do deep imports, but then that code will crash at runtime).

**Comment by JoshuaKGoldberg** [comment]
Ah, you're right - I forgot about this convoluted chain.

tl;dr that I'll copy & paste around:

* We dropped support for Node <=14
* As a part of ☝️ we removed the entries in our `package.json`s that helped old Node versions resolve package entries ("module resolution")
* TypeScript respects Node module resolution algorithms, which means we essentially dropped _types_ module resolution support for older versions of Node
* TSConfig `"moduleResolution": "node"` is actually an alias of `"node10"`, which is EOL - you should switch to `"node16"` or `"nodeNext"` or whatever is most accurate for your project per https://www.typescriptlang.org/tsconfig#moduleResolution

Working as intended.
